File: 3243-shortest-distance-after-road-addition-queries-i/3243-shortest-distance-after-road-addition-queries-i.py
class Solution(object):
    def shortestDistanceAfterQueries(self, n, queries):
        """
        :type n: int
        :type queries: List[List[int]]
        :rtype: List[int]
        """
        neigh = [[] for _ in range(n)]
        
        for i in range(1,n):
            neigh[i].append(i-1)

        inf = 10**9
            
        ans = []
        for [u,v] in queries:
            neigh[v].append(u)
            
            dp = [inf]*(n)
            dp[0] = 0
            for i in range(1,n):
                
                for pre in neigh[i]:
                    dp[i] = min(dp[i], dp[pre] + 1)
                    
            ans.append(dp[n-1])
            
        return ans

File: 3243-shortest-distance-after-road-addition-queries-i/3243-shortest-distance-after-road-addition-queries-i.cpp
class Solution {
public:
    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        int f[505];
        memset(f,127/3,sizeof f);
        vector<int> oo;
        vector<int> L[505];
        for(int i=0;i<queries.size();++i) {
            L[queries[i][1]].push_back(queries[i][0]);
            int pt=0;
            f[0]=0;
            for(int j=i?queries[i][1]:1;j<n;++j) {
                f[j]=min(f[j],f[j-1]+1);
                for(auto l:L[j]) {
                    f[j]=min(f[j],f[l]+1);
                }
            }
            oo.push_back(f[n-1]);
        }
        return oo;
    }
};